src/Sparrow.Server/ByteString.cs

```diff
@@ -841,21 +841,108 @@ public void Reset()
 
                 ReleaseSegment(segment);
             }
-            
+
             _wholeSegments.Clear();
             _wholeSegments.Add(_internalCurrent);
             _wholeSegments.Add(_externalCurrent);
         }
 
+        public void DefragmentSegments()
+        {
+            // small allocators
+            if (_totalAllocated < 128 * Sparrow.Global.Constants.Size.Megabyte)
+                return;
+
+            // small fragmentation
+            var segments = _internalReadyToUseMemorySegments;
+            if (segments == null || segments.Count < 1024)
+                return;
+
+            var newSegments = MergeSegments(segments, ascending: true);
+            newSegments = MergeSegments(newSegments, ascending: false);
+
+            Debug.Assert(_internalReadyToUseMemorySegments.Count == 0);
+            _internalReadyToUseMemorySegments.AddRange(newSegments);
+
+            static SegmentInformation FindSegment(List<SegmentInformation> segments, byte* start)
+            {
+                for (var i = 0; i < segments.Count; i++)
+                {
+                    var otherSegment = segments[i];
+                    if (otherSegment.Start != start)
+                        continue;
+
+                    segments.RemoveAt(i);
+                    return otherSegment;
+                }
+
+                return null;
+            }
+
+            static SegmentInformation FindSegmentBackwards(List<SegmentInformation> segments, byte* start)
+            {
+                for (var i = segments.Count - 1; i > 0; i--)
+                {
+                    var otherSegment = segments[i];
+                    if (otherSegment.Start != start)
+                        continue;
+
+                    segments.RemoveAt(i);
+                    return otherSegment;
+                }
+
+                return null;
+            }
+
+            static List<SegmentInformation> MergeSegments(List<SegmentInformation> segments, bool ascending)
+            {
+                var newSegments = new List<SegmentInformation>();
+
+                while (segments.Count > 0)
+                {
+                    var index = ascending ? 0 : segments.Count - 1;
+
+                    var segment = segments[index];
+                    segments.RemoveAt(index);
+
+                    var segmentSize = segment.Size;
+                    Debug.Assert(segment.Size == segment.SizeLeft);
+
+                    var nextSegmentStart = segment.End;
+
+                    while (true)
+                    {
+                        var nextSegment = ascending
+                            ? FindSegment(segments, nextSegmentStart)
+                            : FindSegmentBackwards(segments, nextSegmentStart);
```

Comment:
When searching backward then `nextSegmentStart` is still the start for the next segment?

Reply:
Segments on the list are not ordered by the Start, so when we are processing them forward, we might (and we do) encounter a situation where we have segment with start 10, and next segment starts with 8, meaning that if we would move backwards then they would get merged. This means that the nextSegmentStart is still correct. This also answers your other question (https://github.com/ravendb/ravendb/pull/15590#discussion_r1066646326)